File: scripts/flag-duplicate-questions.py
"""
Flag near-duplicate published quiz questions within the same topic.

Uses the same sentence-transformers model as the pipeline (all-MiniLM-L6-v2).
For each pair of published questions in a topic with cosine similarity above
the threshold, keeps the higher-confidence question and flags the other.

Usage (from repo root):
  set -a && source .env.local && set +a
  python scripts/flag-duplicate-questions.py [--dry-run] [--threshold 0.82]
"""

import argparse
import os
import sys
import numpy as np
import psycopg2
import psycopg2.extras
from sentence_transformers import SentenceTransformer

parser = argparse.ArgumentParser()
parser.add_argument('--dry-run', action='store_true')
parser.add_argument('--threshold', type=float, default=0.82)
args = parser.parse_args()

THRESHOLD = args.threshold
DRY_RUN   = args.dry_run

print(f"\nDuplicate question auditor")
print(f"  threshold : {THRESHOLD}")
print(f"  dry-run   : {DRY_RUN}\n")

db_url = os.environ.get('DATABASE_URL') or os.environ.get('DIRECT_URL')
if not db_url:
    print("ERROR: DATABASE_URL or DIRECT_URL must be set")
    sys.exit(1)

conn = psycopg2.connect(db_url)
cur  = conn.cursor(cursor_factory=psycopg2.extras.DictCursor)

model = SentenceTransformer('all-MiniLM-L6-v2')

def cosine(a, b):
    return float(np.dot(a, b) / (np.linalg.norm(a) * np.linalg.norm(b) + 1e-12))

# Fetch all topics
cur.execute("""
    SELECT t.id, t.title, s.name as subject_name
    FROM topics t
    JOIN subjects s ON s.id = t.subject_id
    ORDER BY s.name, t.title
""")
topics = cur.fetchall()

total_pairs   = 0
total_flagged = 0

for topic in topics:
    cur.execute("""
        SELECT id, question_text, confidence_score
        FROM quiz_questions
        WHERE topic_id = %s AND status = 'published'
    """, (topic['id'],))
    questions = cur.fetchall()

    if len(questions) < 2:
        continue

    texts = [q['question_text'] or '' for q in questions]
    embeddings = model.encode(texts, normalize_embeddings=True)

    flagged = set()

    for i in range(len(questions)):
        for j in range(i + 1, len(questions)):
            if questions[i]['id'] in flagged or questions[j]['id'] in flagged:
                continue

            sim = cosine(embeddings[i], embeddings[j])
            if sim < THRESHOLD:
                continue

            total_pairs += 1
            qi = questions[i]
            qj = questions[j]
            score_i = float(qi['confidence_score'] or 0)
            score_j = float(qj['confidence_score'] or 0)
            keep   = qi if score_i >= score_j else qj
            to_flag = qj if score_i >= score_j else qi

            print(f"  [{topic['subject_name']}] {topic['title']}")
            print(f"    similarity : {sim:.3f}")
            print(f"    keep  (score={float(keep['confidence_score'] or 0):.1f}) : \"{(keep['question_text'] or '')[:80]}…\"")
            print(f"    flag  (score={float(to_flag['confidence_score'] or 0):.1f}) : \"{(to_flag['question_text'] or '')[:80]}…\"")

            if not DRY_RUN:
                cur.execute(
                    "UPDATE quiz_questions SET status = 'flagged' WHERE id = %s",
                    (to_flag['id'],)
                )
                conn.commit()
                print(f"    ✓ flagged")

            flagged.add(to_flag['id'])
            total_flagged += 1
            print()

conn.close()

verb = "would be" if DRY_RUN else ""
print(f"\nSummary: {total_pairs} duplicate pair(s), {total_flagged} question(s) {verb} flagged.")
if DRY_RUN and total_flagged > 0:
    print("Re-run without --dry-run to apply.")

File: scripts/flag-duplicate-questions.ts
/**
 * Flag near-duplicate published quiz questions within the same topic.
 *
 * Fetches all published questions per topic, embeds them via OpenAI, then
 * computes pairwise cosine similarity. For each pair above the threshold,
 * keeps the higher-confidence question and flags the other as 'flagged'.
 *
 * Usage:
 *   set -a && source .env.local && set +a
 *   npx tsx scripts/flag-duplicate-questions.ts [--dry-run] [--threshold 0.82]
 */

import { prisma } from '../lib/prisma'

const args = process.argv.slice(2)
const DRY_RUN = args.includes('--dry-run')
const thresholdIdx = args.findIndex(a => a === '--threshold')
const THRESHOLD = thresholdIdx >= 0 ? parseFloat(args[thresholdIdx + 1]) : 0.82

console.log(`\nDuplicate question auditor`)
console.log(`  threshold : ${THRESHOLD}`)
console.log(`  dry-run   : ${DRY_RUN}\n`)

function cosine(a: number[], b: number[]): number {
  let dot = 0, na = 0, nb = 0
  for (let i = 0; i < a.length; i++) {
    dot += a[i] * b[i]
    na += a[i] * a[i]
    nb += b[i] * b[i]
  }
  return dot / (Math.sqrt(na) * Math.sqrt(nb) + 1e-12)
}

async function embedBatch(texts: string[]): Promise<number[][]> {
  const res = await fetch('https://api.openai.com/v1/embeddings', {
    method: 'POST',
    headers: {
      'Authorization': `Bearer ${process.env.OPENAI_API_KEY}`,
      'Content-Type': 'application/json',
    },
    body: JSON.stringify({ model: 'text-embedding-ada-002', input: texts }),
  })
  if (!res.ok) throw new Error(`OpenAI embedding error: ${res.status} ${await res.text()}`)
  const json = await res.json() as { data: { embedding: number[] }[] }
  return json.data.map(d => d.embedding)
}

async function main() {
  if (!process.env.OPENAI_API_KEY) {
    console.error('OPENAI_API_KEY not set. Run: set -a && source .env.local && set +a')
    process.exit(1)
  }

  const topics = await prisma.topic.findMany({
    select: { id: true, title: true, subject: { select: { name: true } } },
  })

  let totalPairs = 0
  let totalFlagged = 0

  for (const topic of topics) {
    const subjectName = topic.subject?.name ?? 'unknown'

    const questions = await prisma.quizQuestion.findMany({
      where: { topic_id: topic.id, status: 'published' },
      select: { id: true, question_text: true, confidence_score: true },
    })

    if (questions.length < 2) continue

    // Embed in batches of 20
    const texts = questions.map(q => q.question_text ?? '')
    const embeddings: number[][] = []
    for (let i = 0; i < texts.length; i += 20) {
      const batch = await embedBatch(texts.slice(i, i + 20))
      embeddings.push(...batch)
    }

    const flagged = new Set<string>()

    for (let i = 0; i < questions.length; i++) {
      for (let j = i + 1; j < questions.length; j++) {
        if (flagged.has(questions[i].id) || flagged.has(questions[j].id)) continue
        const sim = cosine(embeddings[i], embeddings[j])
        if (sim < THRESHOLD) continue

        totalPairs++
        const qi = questions[i]
        const qj = questions[j]
        const keepIdx = (Number(qi.confidence_score) ?? 0) >= (Number(qj.confidence_score) ?? 0) ? i : j
        const flagIdx = keepIdx === i ? j : i
        const kept = questions[keepIdx]
        const toFlag = questions[flagIdx]

        console.log(`  [${subjectName}] ${topic.title}`)
        console.log(`    similarity : ${sim.toFixed(3)}`)
        console.log(`    keep  (score=${Number(kept.confidence_score).toFixed(1)}) : "${(kept.question_text ?? '').slice(0, 80)}…"`)
        console.log(`    flag  (score=${Number(toFlag.confidence_score).toFixed(1)}) : "${(toFlag.question_text ?? '').slice(0, 80)}…"`)

        if (!DRY_RUN) {
          await prisma.quizQuestion.update({
            where: { id: toFlag.id },
            data: { status: 'flagged' },
          })
          console.log(`    ✓ flagged`)
        }
        flagged.add(toFlag.id)
        totalFlagged++
        console.log()
      }
    }
  }

  console.log(`\nSummary: ${totalPairs} duplicate pair(s), ${totalFlagged} question(s) ${DRY_RUN ? 'would be' : ''} flagged.`)
  if (DRY_RUN && totalFlagged > 0) console.log('Re-run without --dry-run to apply.')
}

main()
  .catch(err => { console.error(err); process.exit(1) })
  .finally(() => prisma.$disconnect())
